Update deprecated code in CameraXBasic

donovanfm opened the pull request:
This pull request addresses issue https://github.com/android/camera-samples/issues/512.

Main changes:
 - Replace externalMediaDirs with MediaStore
 - Replace onRequestPermissionsResult with registerForActivityResult
 - Replace ListenableFuture with asynchronous coroutine
 - Replace View.systemUiVisibility with WindowInsetsControllerCompat (falling back for api level 29 and before)
 - Replace FragmentStatePagerAdapter with FragmentStateAdapter

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/utils/MediaStoreUtils.kt:
```diff
@@ -0,0 +1,95 @@
+
+    // Suppress DATA index deprecation warning since we need the file location for the Glide library
+    @Suppress("DEPRECATION")
+    private val imageDataColumnIndex = MediaStore.Images.Media.DATA
```
nit: these should remain as `private const val` -- should they be placed within a `private companion object` ?

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/utils/MediaStoreUtils.kt:
```diff
@@ -0,0 +1,95 @@
+        context.getExternalFilesDir(null)?.toUri()
+    }
+
+    private fun getMediaStoreImageCursor(mediaStoreCollection: Uri): Cursor? {
```
Should these async operations be converted to `suspend fun` and switch to the correct dispatcher?
ie.
```
private suspend fun getMediaStoreImageCursor(mediaStoreCollection: Uri): Cursor? =
    withContext(Dispatchers.IO) {
        ...
    }
```
That way it's clear to the callsite that this is an async operation and it internally ensures that the operation is performed on the correct thread
And the result would be returned on the calling thread which could be UI thread which should eliminate doing something like `someUiElement.post { ... }`

donovanfm replied:
I believe I addressed this concern, but please take another look and let me know.

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/utils/MediaStoreUtils.kt:
```diff
@@ -0,0 +1,95 @@
+        val cursor = getMediaStoreImageCursor(mediaStoreCollection)
+        if (cursor?.moveToFirst() != true) return null
+
+        val filename = cursor.getString(cursor.getColumnIndexOrThrow(imageDataColumnIndex))
```
If the operation throws an exception then this means the cursor wouldn't be closed I think
Maybe it's safer to wrap in a try / finally block?

donovanfm replied:
Thanks for catching! I did that, and then Android Studio suggested `use`, which looks even cleaner.

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/utils/MediaStoreUtils.kt:
```diff
@@ -0,0 +1,95 @@
+    }
+}
+
+class MediaStoreFile(val uri: Uri, val file: File, val id: Long)
```
should this be a data class or was it intentional that equals / hashcode / toString wasn't needed?

donovanfm replied:
Just an oversight. Updated.

jsaund reviewed (changes requested):
overall looks good!
left a few comments
also maybe a follow-up but I wonder if rooms may fit in to this instead of manually managing the cursor / content resolver

donovanfm commented:
Thanks for the review, Jag! I addressed your comments, and I can look into Rooms to clean up the cursor code. I agree that it can be a follow-up change.

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt:
```diff
@@ -169,7 +160,7 @@ class CameraFragment : Fragment() {
+    private fun setGalleryThumbnail(filename: String) {
         // Run the operations in the view's thread
         cameraUiContainerBinding?.photoViewButton?.let { photoViewButton ->
             photoViewButton.post {
```
nit: since this should all be main thread safe now -- we can probably remove the `.post` call but just double check this is true

donovanfm replied:
Removing this causes a Glide error, saying it's not called on the UI thread. I could alternatively put this in a `launch` block, but I chose to leave it as is. Let me know if you'd like a followup change for this.

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt:
```diff
@@ -470,10 +458,9 @@ class CameraFragment : Fragment() {
 
         // In the background, load latest photo taken (if any) for gallery thumbnail
         lifecycleScope.launch(Dispatchers.IO) {
```
this can be `lifecycleScope.launch {` now since `getLatestImageFilename` is main thread safe

jsaund commented on CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/GalleryFragment.kt:
```diff
@@ -52,29 +49,28 @@ class GalleryFragment internal constructor() : Fragment() {
-        mediaList = rootDirectory.listFiles { file ->
-            EXTENSION_WHITELIST.contains(file.extension.toUpperCase(Locale.ROOT))
-        }?.sortedDescending()?.toMutableList() ?: mutableListOf()
+        runBlocking {
```
we should avoid `runBlocking` because it's going to block the main thread
maybe
```
private var mediaList: MutableList<MediaStoreFile> = emptyList()
...
lifecycleScope.launch {
    mediaList = MediaStoreUtils(requireContext()).getImages()
    adapter.mediaItems = mediaList
}
```

donovanfm replied:
I originally tried exactly the code you have, but the app crashed in `onViewCreated` when `mediaList` wasn't defined yet. I've tested the app with runBlocking and there is no lag, even with over 100 photos to show in the gallery carousel. The only other option I can think of is async/await, but that would put most of the code in `onViewCreated` in a launch block after the call to `await` anyway. So that seems to have latency issues as well. Still, perhaps that's better because the main thread isn't completely blocked, even though nothing can be shown on the photo gallery until the photos have been loaded. I can implement it that way if you think it's preferable. LMK.

donovanfm replied:
I implemented this with `async` in `onCreate` and the app is working as expected:

```
mediaListDeferred = lifecycleScope.async {
    // Get images this app has access to from MediaStore
    val mediaStoreUtils = MediaStoreUtils(requireContext())
    mediaStoreUtils.getImages()
}
```
Then I put most of the `onViewCreated` code after `mediaList = mediaListDeferred.await()` in a `launch` block. I can add that commit if this approach sounds good to you.

donovanfm replied:
Thanks for the offline conversation on this. I've implemented your suggested changes and removed `runBlocking` completely.

jsaund reviewed (approved):
Left a few suggestions
LGTM overall -- ship when ready 🙂
